pystac/summaries.py

```diff
@@ -0,0 +1,270 @@
+import sys
+import json
+import numbers
+import urllib.request
+from enum import Enum
+
+import pystac
+from pystac.utils import get_required
+
+from typing import (
+    Any,
+    Dict,
+    Generic,
+    List,
+    Optional,
+    Union,
+    cast,
+    TypeVar,
+    Iterable,
+    TYPE_CHECKING,
+)
+
+if sys.version_info >= (3, 8):
+    from typing import Protocol
+else:
+    from typing_extensions import Protocol
+
+if TYPE_CHECKING:
+    from pystac.item import Item as Item_Type
+    from pystac.collection import Collection as Collection_Type
+
+from abc import abstractmethod
+
+
+class Comparable(Protocol):
+    """Protocol for annotating comparable types."""
+
+    @abstractmethod
+    def __lt__(self: "T", x: "T") -> bool:
+        pass
+
+
+T = TypeVar("T", bound=Comparable)
+
+
+class RangeSummary(Generic[T]):
+    def __init__(self, minimum: T, maximum: T):
+        self.minimum = minimum
+        self.maximum = maximum
+
+    def to_dict(self) -> Dict[str, Any]:
+        return {"minimum": self.minimum, "maximum": self.maximum}
+
+    def update_with_value(self, v: T) -> None:
+        self.minimum = min(self.minimum, v)
+        self.maximum = max(self.maximum, v)
+
+    @classmethod
+    def from_dict(cls, d: Dict[str, Any]) -> "RangeSummary[T]":
+        minimum: T = get_required(d.get("minimum"), "RangeSummary", "minimum")
+        maximum: T = get_required(d.get("maximum"), "RangeSummary", "maximum")
+        return cls(minimum=minimum, maximum=maximum)
+
+
+FIELDS_JSON_URL = "https://cdn.jsdelivr.net/npm/@radiantearth/stac-fields/fields.json"
+
+
+class SummaryStrategy(Enum):
+    ARRAY = "v"
+    RANGE = "r"
+    SCHEMA = "s"
+    DONT_SUMMARIZE = False
+    DEFAULT = True
+
+
+class Summarizer:
+    """The Summarizer computes summaries from values, following the definition of fields
+    to summarize provided in a json file.
+
+    For more information about the structure of the fields json file, see:
+
+    https://github.com/stac-utils/stac-fields
+
+    Args:
+        fields (str): the path to the json file with field descriptions.
+        If no file is passed, a default one will be used.
+    """
+
+    _default_field_definitions: Dict[str, Any] = {}
+
+    def __init__(self, fields: Optional[str] = None):
+        if fields is None:
+            self._set_default_field_definitions()
+        else:
+            jsonfields = pystac.StacIO.default().read_json(fields)
+            self._set_field_definitions(jsonfields)
```

Comment:
Since this is going to get called any time a `Summarizer` is instantiated without a `fields` argument, maybe it makes sense to move the logic for fetching the JSON in `_set_default_field_definitions` into a separate function and use `functools.lru_cache` to cache this across all instances. 

It seems like we could re-use that cached function for both of these cases (with or without `fields` argument), which would help with performance in the case where they are providing a custom URL and instantiating the `Summarizer` multiple times.

Something like:

```python
@lru_cache
def _get_fields_json(url: str) -> Dict[str, Any]:
    return pystac.StacIO.default().read_json(url)
```

Reply:
Actually, the case of the default fields definition being downloaded multiple time was already avoided, since the variable storing that was a class variable, and fetching the definition file was only done if that variable was none. But I think that your solution is much better and will also cache non-default files, so I have changed the mechanism.

I have also simplified things a bit, using only the `read_json `method from `StacIO`

Reply:
> Actually, the case of the default fields definition being downloaded multiple time was already avoided, since the variable storing that was a class variable, and fetching the definition file was only done if that variable was none.

Right! Sorry I missed that.